robotoff/prediction/upc_image.py

```diff
@@ -0,0 +1,83 @@
+from enum import Enum
+
+import cv2
+import numpy as np
+
+from robotoff.utils import get_logger
+
+logger = get_logger(__name__)
+
+
+class UPCImageType(Enum):
+    """This enum is used to determine the type of image regarding UPC Information"""
+
+    """The image is a UPC Image which means that a UPC has a greater area in the image than the threshold and thus it is a poor quality main photo"""
+    UPC_IMAGE = "UPC_IMAGE"
+
+    """The image is not a UPC Image but a UPC is present in the image, however it is too small to be considered a UPC Image"""
+    SMALL_UPC = "SMALL_UPC"
+
+    """The image is not a UPC Image and no UPC is present in the image"""
+    NO_UPC = "NO_UPC"
+
+
+def get_area(box: np.ndarray) -> float:
+    """This method determines the area of a polygon given a list of points
+    The formula for determining area of the points is the shoelace formula https://en.wikipedia.org/wiki/Shoelace_formula
+    For additional info look at https://stackoverflow.com/questions/24467972/calculate-area-of-polygon-given-x-y-coordinates
+    :param box: list of points in format [[y, x], ...] (opencvformat)
+    :return: float of the area of the polygon that is our bounding box
+    """
+    x = np.array([pt[1] for pt in box])
+    y = np.array([pt[0] for pt in box])
+    x, y
+    return 0.5 * np.abs(np.dot(x, np.roll(y, 1)) - np.dot(y, np.roll(x, 1)))
+
+
+def get_is_upc_image_from_image(
+    img: np.ndarray,
+) -> tuple[float, UPCImageType, np.ndarray]:
+    """
+    This method determines if an image is a UPC_Image or not
+    A UPC_Image is defined as an image that has a UPC with a high pct area in the image and thus it is a poor quality main photo
+    :param img: numpy array of the image
+    :return: Enum of the prediction class as a UPCImageType
+    """
+    AREA_THRESHOLD = 0.09  # Best result thus far in testing, a UPC with a greater area in the image means it is a "UPC_Image"
```

Comment:
This threshold looks low to me though, if only 9% of the image is covered by a UPC you consider it an UPC image?

Reply:
I felt similarly when first writing the script but I found it is sneaky how much 9% of the image really is. Happy to tweak this now or if we discover any issues during hunger games but from my testing we were missing some I felt were UPC_Images with a higher threshold. I'm also happy to test on additional data if you want or we can raise it but that was why its only 9%.